services/hierarchy/src/main/kotlin/VulnerabilityService.kt

```diff
@@ -139,10 +141,44 @@ class VulnerabilityService(private val db: Database) {
             .groupBy(VulnerabilitiesTable.id, IdentifiersTable.id)
             .count()
     }
+
+    /** Count vulnerabilities by rating in provided ORT runs. */
+    suspend fun countByRatingForOrtRunIds(vararg ortRunIds: Long): CountByCategory<VulnerabilityRating> = db.dbQuery {
+        val countAlias = Count(stringLiteral("*")).alias("count")
+        val ratingAlias = ratingAlias()
+
+        val ratingToCountMap = mutableMapOf<VulnerabilityRating, Long>().apply {
+            VulnerabilityRating.entries.forEach { this[it] = 0 }
+        }
+
+        val subQuery = VulnerabilitiesTable
+            .innerJoin(AdvisorResultsVulnerabilitiesTable)
+            .innerJoin(AdvisorResultsTable)
+            .innerJoin(AdvisorRunsIdentifiersTable)
+            .innerJoin(AdvisorRunsTable)
+            .innerJoin(AdvisorJobsTable)
+            .innerJoin(IdentifiersTable)
+            .innerJoin(VulnerabilityReferencesTable)
+            .select(VulnerabilitiesTable.id, ratingAlias)
+            .where { AdvisorJobsTable.ortRunId inList ortRunIds.asList() }
+            .groupBy(VulnerabilitiesTable.id, IdentifiersTable.id)
+            .alias("subQuery")
+
+        val rating = subQuery[ratingAlias]
+
+        subQuery
+            .select(rating, countAlias)
+            .groupBy(rating)
+            .forEach { row ->
+                ratingToCountMap[VulnerabilityRating.entries[(row[rating] ?: 0)]] = row[countAlias]
```

Comment:
Just double-checking: The `[]`-operator is an alias for `get()` which seems to throw instead of returning `null` for absent keys. Can someone confirm / dissent?

Reply:
Note that I did not check whether tests cover that case.

Reply:
Yes, using `get()`, the editor gives me a hint _"Should be replaced with indexing"_

Reply:
I can't really produce a test that would produce an out-of-bounce rating, as the ratingAlias should only give an overall rating within those bounds

Reply:
Right, but do `[]` / `get()` throw or return `null` for missing keys?

Reply:
Please see

```
class ResultRow(
    /** Mapping of the expressions stored on this row to their index positions. */
    val fieldIndex: Map<Expression<*>, Int>,
    private val data: Array<Any?> = arrayOfNulls<Any?>(fieldIndex.size)
) {
    private val database: Database? = TransactionManager.currentOrNull()?.db

    private val lookUpCache = ResultRowCache()

    /**
     * Retrieves the value of a given expression on this row.
     *
     * @param expression expression to evaluate
     * @throws IllegalStateException if expression is not in record set or if result value is uninitialized
     *
     * @see [getOrNull] to get null in the cases an exception would be thrown
     */
    @Suppress("UNCHECKED_CAST")
    operator fun <T> get(expression: Expression<T>): T {
```

So I guess you should use `getOrNull()` instead.